core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala

```diff
@@ -1100,6 +1100,64 @@ class ReplicaFetcherThreadTest {
     assertEquals(Collections.singletonList(tid1p0), fetchRequestBuilder2.removed())
   }
 
+  @Test
+  def testFetchResponseWithPartitionData(): Unit = {
+    verifyLocalFetchCompletionAfterHighWatermarkUpdate(hasHighWatermarkChanged = true)
+    verifyLocalFetchCompletionAfterHighWatermarkUpdate(hasHighWatermarkChanged = false)
+  }
+
+  private def verifyLocalFetchCompletionAfterHighWatermarkUpdate(hasHighWatermarkChanged: Boolean): Unit = {
+    val props = TestUtils.createBrokerConfig(1, "localhost:1234")
+    val config = KafkaConfig.fromProps(props)
+    val newHighWatermark = 100L
+
+    val mockBlockingSend: BlockingSend = mock(classOf[BlockingSend])
+    when(mockBlockingSend.brokerEndPoint()).thenReturn(brokerEndPoint)
+
+    val log: UnifiedLog = mock(classOf[UnifiedLog])
+    when(log.updateHighWatermark(newHighWatermark))
+      .thenReturn(HighWatermarkUpdate(newHighWatermark, hasHighWatermarkChanged))
+
+    val appendInfo: Option[LogAppendInfo] = Some(mock(classOf[LogAppendInfo]))
```

Comment:
`val appendInfo = Partition.appendRecordstoFollowerOrFutureReplica()` can only be None when isFuture = true, i.e. in `ReplicaAlterLogDirsThread.processPartitionData()`. removed the test case for it